### PR TITLE
Implement update loader and log loaded application info

### PR DESCRIPTION
**Pull request by Xpl0itR**
image

image

### CODE REVIEW

**Comment by ReaperOfSouls1909** [comment]
```HOS\Horizon.cs(487,80): error CS1503: Argument 2: cannot convert from 'string' to 'LibHac.Common.U8Span' [C:\projects\ryujinx-mainline-test\Ryujinx.HLE\Ryujinx.HLE.csproj]
HOS\Horizon.cs(499,61): error CS1503: Argument 2: cannot convert from 'string' to 'LibHac.Common.U8Span' [C:\projects\ryujinx-mainline-test\Ryujinx.HLE\Ryujinx.HLE.csproj]``` when trying to compile once that error is fixed i will get better feedback

**Comment by LePwnerer** [comment]
image
Suggestion: Add Game Title & Current Version abov "Available Updates"

**Comment by ReaperOfSouls1909** [comment]
After all my tests I found no issues with this PR I've tested any game that would load an update and even some that had 7 updates great work as for feedback ++ when you add the update have it default to that new update.

**Comment by legend800** [comment]
Similar experience here. Tried several TUs and had no issues. Great job, been waiting for this!

**Comment by AcK77** [comment]
It could be nice to get the title version near the title id in the window titlebar.
Additionally we could add the title id and the title version in the logging too.
Because if we don’t add any informations it will be a mess to debug things later.

**Comment by Xpl0itR** [comment]
Updated

**Comment by MelonSpeedruns** [comment]
LGTM!

**Comment by legend800** [comment]
Quick note: I installed the latest build from the pr yesterday, selected latest Mario Kart 8 TU, started game, save file (that I've had for a month+ with many builds swaps) got wiped. Even with the 1.0 selected it was still gone. 

Could just be a coincidence however. I don't want to add a false positive here, so if this PR could not have caused that and it's completely unrelated, disregard.

File: Ryujinx.HLE/HOS/Horizon.cs [hunk 1]
```diff
@@ -455,6 +464,51 @@ public void LoadNca(Nca mainNca, Nca patchNca, Nca controlNca)
...
+                            }
+                        }
+
+                        foreach (DirectoryEntryEx fileEntry in nsp.EnumerateEntries("/", "*.nca"))
+                        {
+                            nsp.OpenFile(out IFile ncaFile, fileEntry.FullPath.ToU8Span(), OpenMode.Read).ThrowIfFailure();
+
+                            Nca nca = new Nca(KeySet, ncaFile.AsStorage());
+
+                            if ($"{nca.Header.TitleId.ToString("x16")[..^3]}000" != mainNca.Header.TitleId.ToString("x16")) break;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Horizon.cs:497]: Nit: Expand brackets

File: Ryujinx/Ui/ApplicationLibrary.cs [hunk 2]
```diff
@@ -623,5 +624,65 @@ private static void GetNameIdDeveloper(Nacp controlData, out string titleName, o
...
+            {
+                using (Stream stream = File.OpenRead(jsonPath))
+                {
+                    IJsonFormatterResolver resolver = CompositeResolver.Create(StandardResolver.AllowPrivateSnakeCase);
+                    string updatePath = JsonSerializer.Deserialize<TitleUpdateMetadata>(stream, resolver).Selected;
+
+                    if (!File.Exists(updatePath))
+                    {
+                        version = "";
+                        return false;
```
Comment (AcK77) [inline Ryujinx/Ui/ApplicationLibrary.cs:642]: New empty line before a return

File: Ryujinx/Ui/ApplicationLibrary.cs [hunk 3]
```diff
@@ -623,5 +624,65 @@ private static void GetNameIdDeveloper(Nacp controlData, out string titleName, o
...
+
+                            if (nca.Header.ContentType == NcaContentType.Control)
+                            {
+                                BlitStruct<ApplicationControlProperty> controlData = new BlitStruct<ApplicationControlProperty>(1);
+
+                                nca.OpenFileSystem(NcaSectionType.Data, IntegrityCheckLevel.None).OpenFile(out IFile nacpFile, "/control.nacp".ToU8Span(), OpenMode.Read).ThrowIfFailure();
+                                nacpFile.Read(out long _, 0, controlData.ByteSpan, ReadOption.None).ThrowIfFailure();
+
+                                version = controlData.Value.DisplayVersion.ToString();
+                                return true;
```
Comment (AcK77) [inline Ryujinx/Ui/ApplicationLibrary.cs:677]: New empty line before a return

File: Ryujinx/Ui/ApplicationLibrary.cs [hunk 4]
```diff
@@ -623,5 +624,65 @@ private static void GetNameIdDeveloper(Nacp controlData, out string titleName, o
...
+                            }
+                        }
+                    }
+                }
+            }
+
+            version = "";
+            return false;
```
Comment (AcK77) [inline Ryujinx/Ui/ApplicationLibrary.cs:685]: New empty line before a return

File: Ryujinx/Ui/ApplicationLibrary.cs [hunk 5]
```diff
@@ -623,5 +624,65 @@ private static void GetNameIdDeveloper(Nacp controlData, out string titleName, o
...
+                            }
+                        }
+
+                        foreach (DirectoryEntryEx fileEntry in nsp.EnumerateEntries("/", "*.nca"))
+                        {
+                            nsp.OpenFile(out IFile ncaFile, fileEntry.FullPath.ToU8Span(), OpenMode.Read).ThrowIfFailure();
+
+                            Nca nca = new Nca(_virtualFileSystem.KeySet, ncaFile.AsStorage());
+
+                            if ($"{nca.Header.TitleId.ToString("x16")[..^3]}000" != titleId) break;
```
Comment (AcK77) [inline Ryujinx/Ui/ApplicationLibrary.cs:667]: Expand brackets

File: Ryujinx/Ui/TitleUpdateWindow.cs [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+using Ryujinx.HLE.FileSystem;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using LibHac;
+using LibHac.Common;
+using LibHac.Ns;
+using LibHac.Spl;
+using Utf8Json;
```
Comment (AcK77) [inline Ryujinx/Ui/TitleUpdateWindow.cs:16]: Sort using

File: Ryujinx/Ui/TitleUpdateWindow.cs [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+            foreach (string path in _titleUpdateWindowData.Paths)
+            {
+                AddUpdate(path);
+            }
+
+            _noUpdateRadioButton.Active = true;
+            foreach (KeyValuePair<RadioButton, string> keyValuePair in _radioButtonToPathDictionary)
+            {
+                if (keyValuePair.Value == _titleUpdateWindowData.Selected)
```
Comment (AcK77) [inline Ryujinx/Ui/TitleUpdateWindow.cs:76]: Add brackets

**Comment by Xpl0itR** [comment]
👍 done

File: Ryujinx/Ui/ApplicationLibrary.cs [hunk 8]
```diff
@@ -629,5 +630,71 @@ private static void GetNameIdDeveloper(ref ApplicationControlProperty controlDat
...
+                            Nca nca = new Nca(_virtualFileSystem.KeySet, ncaFile.AsStorage());
+
+                            if ($"{nca.Header.TitleId.ToString("x16")[..^3]}000" != titleId)
+                            {
+                                break;
+                            }
+
+                            if (nca.Header.ContentType == NcaContentType.Control)
+                            {
+                                BlitStruct<ApplicationControlProperty> controlData = new BlitStruct<ApplicationControlProperty>(1);
```
Comment (Thealexbarney) [inline Ryujinx/Ui/ApplicationLibrary.cs:681]: BlitStruct isn't needed here. You could do `ApplicationControlProperty controlData = new ApplicationControlProperty();` with `SpanHelpers.AsByteSpan()` when doing the `Read()` below.

File: Ryujinx/Ui/TitleUpdateWindow.cs [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+                    {
+                        nsp.OpenFile(out IFile ncaFile, fileEntry.FullPath.ToU8Span(), OpenMode.Read).ThrowIfFailure();
+
+                        Nca nca = new Nca(_virtualFileSystem.KeySet, ncaFile.AsStorage());
+
+                        if ($"{nca.Header.TitleId.ToString("x16")[..^3]}000" == _titleId)
+                        {
+                            if (nca.Header.ContentType == NcaContentType.Control)
+                            {
+                                BlitStruct<ApplicationControlProperty> controlData = new BlitStruct<ApplicationControlProperty>(1);
```
Comment (Thealexbarney) [inline Ryujinx/Ui/TitleUpdateWindow.cs:113]: BlitStruct isn't needed here. You could do `ApplicationControlProperty controlData = new ApplicationControlProperty();` with `SpanHelpers.AsByteSpan()` when doing the `Read()` below.

**Comment by Xpl0itR** [comment]
Fixed

**Pull request merged**